### PR TITLE
Fix #13044: [BUG] NullReferenceException in Azure.Identity 1.2.0-preview.4

### LINKED ISSUE #13044: [BUG] NullReferenceException in Azure.Identity 1.2.0-preview.4
**Describe the bug**
Upgrading from 1.1 authentication fails with `az` CLI and/or MSI credentials.
The same MSI identity between 1.1 and 1.2.
Reverting back to 1.1 solved the issue.

**Expected behavior**
It should keep working.

**Actual behavior (include Exception or Stack Trace)**
In my case access to AppSettings fails on an update attempt - sometimes, but seems to succeed on reads. The relevant frame of the stack trace:
```
Inner exception System.NullReferenceException handled at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw:
   at Azure.Identity.DefaultAzureCredential+<GetTokenFromSourcesAsync>d__12.MoveNext (Azure.Identity, Version=1.2.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Azure.Identity.DefaultAzureCredential+<GetTokenImplAsync>d__11.MoveNext (Azure.Identity, Version=1.2.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8)
```

**To Reproduce**
Please ping me on Teams for repo/sample.

**Environment:**
 - Name and version of the Library package used: Azure.Identity 1.2.0-preview.4 
 - Hosting platform or OS and .NET runtime version: Functions V3, .NET SDK 3.1.300
 - IDE and version : N/A

### PR DESCRIPTION
**Pull request by AlexanderSher**

### CODE REVIEW

File: sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs [hunk 1]
```diff
@@ -102,7 +103,7 @@ public async Task ValidateSharedTokenCacheCredentialSucceededEvents()
         {
             var mockMsalClient = new MockMsalPublicClient
             {
-                Accounts = new IAccount[] { new MockAccount("mockuser@mockdomain.com") },
+                Accounts = new List<IAccount> { new MockAccount("mockuser@mockdomain.com") },
```
Comment (schaabs) [inline sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs:106]: Is this change needed?

File: sdk/identity/Azure.Identity/src/MsalClientBase.cs [hunk 2]
```diff
@@ -11,7 +12,7 @@ namespace Azure.Identity
     internal abstract class MsalClientBase<TClient>
         where TClient : IClientApplicationBase
     {
-        private readonly Lazy<Task> _ensureInitAsync;
+        private readonly AsyncLockWithValue<TClient> _clientAsyncLock;
```
Comment (schaabs) [inline sdk/identity/Azure.Identity/src/MsalClientBase.cs:15]: I'm a bit worried about refactoring classes which are unrelated to this fix so close to release.

File: sdk/identity/Azure.Identity/src/MsalConfidentialClient.cs [hunk 3]
```diff
@@ -46,20 +46,17 @@ protected override async Task<IConfidentialClientApplication> CreateClientAsync(
...
         }
 
-        public virtual async Task<AuthenticationResult> AcquireTokenForClientAsync(string[] scopes, bool async, CancellationToken cancellationToken)
+        public virtual async ValueTask<AuthenticationResult> AcquireTokenForClientAsync(string[] scopes, bool async, CancellationToken cancellationToken)
         {
-            await EnsureInitializedAsync(async).ConfigureAwait(false);
-
-            return await Client.AcquireTokenForClient(scopes).ExecuteAsync(async, cancellationToken).ConfigureAwait(false);
+            IConfidentialClientApplication client = await GetClientAsync(async, cancellationToken).ConfigureAwait(false);
+            return await client.AcquireTokenForClient(scopes).ExecuteAsync(async, cancellationToken).ConfigureAwait(false);
```
Comment (schaabs) [inline sdk/identity/Azure.Identity/src/MsalConfidentialClient.cs:59]: Won't this always result in a `Task` being created since it calls MSAL which is always async? If this is the case is anything gained by updating this method to return `ValueTask`?

File: sdk/identity/Azure.Identity/tests/SharedTokenCacheCredentialTests.cs [hunk 4]
```diff
@@ -35,7 +36,7 @@ public async Task VerifyAuthenticationRecordOption()
 
             var mockMsalClient = new MockMsalPublicClient
             {
-                Accounts = new IAccount[] { new MockAccount("nonexpecteduser@mockdomain.com") },
+                Accounts = new List<IAccount> { new MockAccount("nonexpecteduser@mockdomain.com") },
```
Comment (schaabs) [inline sdk/identity/Azure.Identity/tests/SharedTokenCacheCredentialTests.cs:39]: Why are we refactoring these to `List` instead or array?

File: sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs:106]: This is for `Accounts` property to return `List<T>` rather than `IEnumerable<T>`.

File: sdk/identity/Azure.Identity/src/MsalClientBase.cs (hunk 2 above)
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/src/MsalClientBase.cs:15]: That is actually for the #13201. I didn't see that it is moved to the September release, so I can revert it.

File: sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs:106]: I'll revert this change cause it is part of the fix for #13201

File: sdk/identity/Azure.Identity/src/MsalConfidentialClient.cs (hunk 3 above)
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/src/MsalConfidentialClient.cs:59]: `ValueTask` is generally better for async/sync mixed scenario cause it doesn't create expensive `Task` objects when they aren't needed.

File: sdk/identity/Azure.Identity/src/MsalClientBase.cs (hunk 2 above)
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/src/MsalClientBase.cs]: Reverted

File: sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/tests/AzureIdentityEventSourceTests.cs]: Reverted

File: sdk/identity/Azure.Identity/tests/SharedTokenCacheCredentialTests.cs (hunk 4 above)
Comment (AlexanderSher) [reply sdk/identity/Azure.Identity/tests/SharedTokenCacheCredentialTests.cs]: Reverted

**Review by schaabs — APPROVED** [review]
